src/category_theory/limits/shapes/constructions/limits_of_products_and_equalizers.lean

```diff
@@ -1,139 +1,191 @@
-@[simp] def cones_inv : F.cones ⟶ (diagram F).cones :=
-{ app := λ X c,
+variable {i}
+/--
+(Implementation) Show the cone constructed in `build_limit` is limiting, provided the cones used in
+its construction are.
+-/
+def built_is_limit (t₁ : is_limit c₁) (t₂ : is_limit c₂) (hi : is_limit i) :
+  is_limit (build_limit s t hs ht i) :=
+{ lift := λ q,
   begin
-    refine (fork.of_ι _ _).π,
-    { exact pi.lift c.app },
-    { ext ⟨⟨A,B⟩,f⟩,
-      dsimp,
-      simp only [limit.lift_π, limit.lift_π_assoc, fan.mk_π_app, category.assoc],
-      rw ←(c.naturality f),
-      dsimp,
-      simp only [category.id_comp], }
+    refine hi.lift (fork.of_ι _ _),
+    { refine t₁.lift (fan.mk _ (λ j, _)),
+      apply q.π.app j },
+    { apply t₂.hom_ext,
+      simp [hs, ht] },
   end,
-  naturality' := λ X Y f, by { ext c j, cases j; tidy, } }.
-
-/-- The natural isomorphism between cones over the
-walking pair diagram `diagram F` and cones over the original diagram `F`. -/
-def cones_iso : (diagram F).cones ≅ F.cones :=
-{ hom := cones_hom F,
-  inv := cones_inv F,
-  hom_inv_id' :=
-  begin
-    ext X c j,
-    cases j,
-    { ext, simp },
-    { ext,
-      have t := c.naturality walking_parallel_pair_hom.left,
-      conv at t { dsimp, to_lhs, simp only [category.id_comp] },
-      simp [t], }
-  end }
+  uniq' := λ q m w, hi.hom_ext (i.equalizer_ext (t₁.hom_ext (by simpa using w))) }
 
 end has_limit_of_has_products_of_has_equalizers
 
 open has_limit_of_has_products_of_has_equalizers
 
+/--
+Given the existence of the appropriate (possibly finite) products and equalizers, we know a limit of
+`F` exists.
+(This assumes the existence of all equalizers, which is technically stronger than needed.)
+-/
+lemma has_limit_of_equalizer_and_product (F : J ⥤ C)
+  [has_limit (discrete.functor F.obj)]
+  [has_limit (discrete.functor (λ f : (Σ p : J × J, p.1 ⟶ p.2), F.obj f.1.2))]
+  [has_equalizers C] : has_limit F :=
+has_limit.mk
+{ cone := _,
+  is_limit :=
+    built_is_limit
+      (pi.lift (λ f, limit.π _ _ ≫ F.map f.2))
+      (pi.lift (λ f, limit.π _ f.1.2))
+      (by simp)
+      (by simp)
+      (limit.is_limit _)
+      (limit.is_limit _)
+      (limit.is_limit _) }
+
 /--
 Any category with products and equalizers has all limits.
 
 See https://stacks.math.columbia.edu/tag/002N.
 -/
--- This is not an instance, as it is not always how one wants to construct limits!
 lemma limits_from_equalizers_and_products
   [has_products C] [has_equalizers C] : has_limits C :=
-{ has_limits_of_shape := λ J 𝒥, by exactI
-  { has_limit := λ F, has_limit.of_cones_iso (diagram F) F (cones_iso F) } }
+{ has_limits_of_shape := λ J 𝒥,
+  { has_limit := λ F, by exactI has_limit_of_equalizer_and_product F } }
 
 /--
 Any category with finite products and equalizers has all finite limits.
 
 See https://stacks.math.columbia.edu/tag/002O.
-(We do not prove equivalence with the third condition.)
+(We do not prove equivalence with the third conditionin this file.)
```

Comment:
```suggestion
(We do not prove equivalence with the third condition in this file.)
```

Reply:
What does "this file" refer to in that sentence? I interpret it as pointing to stacks... but I think you could be a bit more explicit.

Reply:
Oops, I meant that the third condition of stacks isn't proved in this file, I'll clarify

Reply:
Actually maybe I'll just remove it altogether, it's not really adding anything useful